software/lamp-os/src/expressions/breathing_expression.cpp

```diff
@@ -0,0 +1,153 @@
+#include "./breathing_expression.hpp"
+
+#include <Arduino.h>
+#include <algorithm>
+#include <cmath>
+
+#include "../util/fade.hpp"
+
+namespace lamp {
+
+// Large frame count to keep animation running continuously
+static constexpr uint32_t BREATHING_MAX_FRAMES = 100000;
+
+BreathingExpression::BreathingExpression(FrameBuffer* inBuffer, uint32_t inFrames)
+    : Expression(inBuffer, inFrames) {
+  isExclusive = false;  // This can run and blend with other things
+  allowedInHomeMode = true;
+}
+
+void BreathingExpression::configureFromParameters(const std::map<std::string, uint32_t>& parameters) {
+  // Extract breath speed parameter (in seconds, convert to ms)
+  auto itSpeed = parameters.find("breathSpeed");
+  uint32_t breathSpeed = (itSpeed != parameters.end()) ? itSpeed->second : 10;
+  breathSpeedMs = breathSpeed * 1000;
+
+  // Color change intervals are set by base Expression::configure() in intervalMinMs/MaxMs
+  // No need to read them from parameters
+
+  // Default to white if no colors provided
+  if (colors.empty()) {
+    colors.push_back(Color(255, 255, 255, 255));
+  }
+
+  // Start with first color
+  targetColor = colors[0];
+}
+
+void BreathingExpression::updateBreathPhase() {
+  uint32_t currentMs = millis();
+
+  if (lastBreathUpdateMs == 0) {
+    lastBreathUpdateMs = currentMs;
+    return;
+  }
+
+  uint32_t deltaMs = currentMs - lastBreathUpdateMs;
+
+  // Update phase based on breath speed
+  // breathSpeedMs is the total cycle time, so increment is deltaMs / breathSpeedMs
+  float phaseIncrement = static_cast<float>(deltaMs) / static_cast<float>(breathSpeedMs);
+  breathPhase += phaseIncrement;
+
+  // Wrap phase to stay in 0-1 range and advance color when cycle completes
+  if (breathPhase >= 1.0f) {
+    breathPhase -= 1.0f;
+
+    // Advance to next color if we have multiple colors
+    if (colors.size() > 1) {
+      if (cyclingForward) {
+        currentColorIndex++;
+        // If we reached the last color, switch to backward
+        if (currentColorIndex >= colors.size() - 1) {
+          currentColorIndex = colors.size() - 1;
+          cyclingForward = false;
+        }
+      } else {
+        // Going backward
+        if (currentColorIndex == 0) {
+          // At first color, switch to forward
+          cyclingForward = true;
+          currentColorIndex = 1;
+        } else {
+          currentColorIndex--;
+        }
+      }
+
+      // Update target color
+      targetColor = colors[currentColorIndex];
+    }
+  }
+
+  lastBreathUpdateMs = currentMs;
+}
+
+void BreathingExpression::onTrigger() {
+  // Save base colors
+  saveBufferState();
```

Comment:
This pattern seems to be a source of weird bugs that I've seen in pulse where it'll randomly snap back to the base colors. Probably best to avoid making additional copies of the buffer when fading